File: packages/davinci/android/src/main/java/com/pingidentity/rndavinci/collector/DaVinciCollectorValueApplier.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rndavinci.collector

import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.pingidentity.davinci.collector.Device
import com.pingidentity.davinci.collector.DeviceAuthenticationCollector
import com.pingidentity.davinci.collector.DeviceRegistrationCollector
import com.pingidentity.davinci.collector.FlowCollector
import com.pingidentity.davinci.collector.MultiSelectCollector
import com.pingidentity.davinci.collector.PhoneNumberCollector
import com.pingidentity.davinci.plugin.Collector
import com.pingidentity.orchestrate.ContinueNode

/**
 * Applies collector value mutations from JS payloads to active DaVinci collectors.
 *
 * DaVinci collectors are identified by key (via `collector.id()`), not by type index.
 */
internal object DaVinciCollectorValueApplier {

    /**
     * Result indicating whether the applied collector triggers an immediate form submit.
     *
     * @property isFlowTrigger True when a FlowCollector key was the only entry in the input.
     */
    data class ApplyResult(val isFlowTrigger: Boolean)

    /**
     * Apply collector values from a JS input map to the current continue node.
     *
     * The input format matches `DaVinciNextInput`:
     * `{ collectors: [{ key: string, value: unknown }] }`
     *
     * Notes:
     * Intentionally non-`suspend` — unlike Journey's equivalent applier. Every DaVinci
     * collector setter in the 2.0.1 SDK (`SingleValueCollector.value`,
     * `MultiSelectCollector.value`, `PhoneNumberCollector.countryCode`/`phoneNumber`,
     * `DeviceRegistrationCollector.value`, `DeviceAuthenticationCollector.value`) is a plain
     * `var` assignment with no async work; suspend overhead is unnecessary. The only
     * suspend work in DaVinci's collector lifecycle happens inside `ContinueNode.next()`
     * (validation + network), which the caller invokes after this function returns.
     *
     * @param node Active continue node providing the collector instances.
     * @param input Bridge input map.
     * @return Result indicating whether an immediate-submit flow was triggered.
     * @throws IllegalArgumentException when a collector key is not found on the node.
     */
    fun apply(node: ContinueNode, input: ReadableMap): ApplyResult {
        val collectorsArray = input.getArray("collectors") ?: return ApplyResult(isFlowTrigger = false)

        val collectorsByKey = node.actions
            .filterIsInstance<Collector<*>>()
            .associateBy { it.id() }

        var isFlowTrigger = false

        for (index in 0 until collectorsArray.size()) {
            val entry = collectorsArray.getMap(index)
                ?: throw IllegalArgumentException("Invalid collector entry at index $index")
            val key = entry.getString("key")
                ?: throw IllegalArgumentException("Missing 'key' in collector entry at index $index")
            val value = readDynamicValue(entry, "value")

            val collector = collectorsByKey[key]
                ?: throw IllegalArgumentException(
                    "No active collector found for key='$key'"
                )

            applyValue(collector, key, value)

            if (collector is FlowCollector) {
                isFlowTrigger = true
            }
        }

        return ApplyResult(isFlowTrigger = isFlowTrigger)
    }

    /**
     * Apply a single value to a collector instance.
     *
     * @param collector Target collector.
     * @param key Collector key (for error messages).
     * @param value Value from JS.
     * @throws IllegalArgumentException when the value type is incompatible with the collector.
     * @throws UnsupportedOperationException when the collector type is not supported by the bridge.
     */
    private fun applyValue(collector: Collector<*>, key: String, value: Any?) {
        when (collector) {
            is MultiSelectCollector -> {
                val list = asStringList(value, key)
                collector.value = list.toMutableList()
            }
            is PhoneNumberCollector -> {
                val map = asStringMap(value)
                map["countryCode"]?.let { collector.countryCode = it }
                map["phoneNumber"]?.let { collector.phoneNumber = it }
            }
            is DeviceRegistrationCollector -> {
                val map = asStringMap(value)
                val deviceType = map["type"]
                    ?: throw IllegalArgumentException("DeviceRegistrationCollector key='$key': value map must include 'type'")
                val device = collector.devices.firstOrNull { it.type == deviceType }
                    ?: throw IllegalArgumentException(
                        "DeviceRegistrationCollector key='$key': no device found with type='$deviceType'"
                    )
                collector.value = device
            }
            is DeviceAuthenticationCollector -> {
                val map = asStringMap(value)
                val deviceType = map["type"]
                    ?: throw IllegalArgumentException("DeviceAuthenticationCollector key='$key': value map must include 'type'")
                val device = collector.devices.firstOrNull { it.type == deviceType }
                    ?: Device(
                        id = map["id"],
                        type = deviceType,
                        title = map["title"] ?: deviceType,
                        description = map["description"] ?: "",
                        iconSrc = map["iconSrc"] ?: "",
                        default = false
                    )
                collector.value = device
            }
            else -> {
                // All remaining collectors (TextCollector, PasswordCollector, SingleValueCollector,
                // SubmitCollector, FlowCollector) extend SingleValueCollector which has setValue(String).
                val strCollector = collector as? com.pingidentity.davinci.collector.SingleValueCollector
                    ?: throw UnsupportedOperationException(
                        "Collector key='$key' type=${collector::class.java.simpleName} is not supported by the bridge"
                    )
                strCollector.value = asString(value, key)
            }
        }
    }

    /**
     * Reads one dynamic value from a bridge map using runtime type inspection.
     */
    private fun readDynamicValue(map: ReadableMap, key: String): Any? {
        if (!map.hasKey(key) || map.isNull(key)) return null
        return when (map.getType(key)) {
            ReadableType.Null -> null
            ReadableType.Boolean -> map.getBoolean(key)
            ReadableType.Number -> normalizeBridgeNumber(map.getDouble(key))
            ReadableType.String -> map.getString(key)
            ReadableType.Map -> map.getMap(key)?.toHashMap()
            ReadableType.Array -> {
                val array = map.getArray(key) ?: return null
                (0 until array.size()).map { index ->
                    when (array.getType(index)) {
                        ReadableType.Null -> null
                        ReadableType.Boolean -> array.getBoolean(index)
                        ReadableType.Number -> normalizeBridgeNumber(array.getDouble(index))
                        ReadableType.String -> array.getString(index)
                        ReadableType.Map -> array.getMap(index)?.toHashMap()
                        ReadableType.Array -> array.getArray(index)
                    }
                }
            }
        }
    }

    /**
     * Normalises a JS number for string coercion parity with iOS.
     *
     * React Native bridges every JS number as a `Double`. An integer-valued double like
     * `5.0` would stringify to `"5.0"` via `Double.toString`, while iOS `NSNumber.stringValue`
     * emits `"5"` for the same input. Collapsing integer-valued doubles to `Long` aligns the
     * two bridges. `NaN` and `+/-Infinity` fail the equality check (NaN is never equal to
     * anything, and `Long.MAX_VALUE.toDouble()` does not equal `Double.POSITIVE_INFINITY`),
     * so they fall through to `Double` and stringify as `"NaN"`/`"Infinity"`.
     *
     * @param bridgedNumber The JS number bridged as a Kotlin Double.
     * @return A `Long` when the value is an integer-valued finite double, otherwise the
     *   original `Double`.
     */
    private fun normalizeBridgeNumber(bridgedNumber: Double): Number {
        val truncatedToLong = bridgedNumber.toLong()
        val isIntegerValued = bridgedNumber == truncatedToLong.toDouble()
        return if (isIntegerValued) truncatedToLong else bridgedNumber
    }

    private fun asString(value: Any?, fieldName: String): String {
        return when (value) {
            is String -> value
            is Number -> value.toString()
            is Boolean -> value.toString()
            else -> throw IllegalArgumentException("$fieldName expects a string-compatible value, got: ${value?.javaClass?.simpleName}")
        }
    }

    private fun asStringList(value: Any?, fieldName: String): List<String> {
        return when (value) {
            is List<*> -> value.mapNotNull { it as? String }
            else -> throw IllegalArgumentException("$fieldName expects an array of strings")
        }
    }

    private fun asStringMap(value: Any?): Map<String, String> {
        val rawMap = value as? Map<*, *> ?: return emptyMap()
        return rawMap.entries.mapNotNull { (rawKey, rawValue) ->
            val stringKey = rawKey as? String ?: return@mapNotNull null
            val stringValue = rawValue?.toString() ?: return@mapNotNull null
            stringKey to stringValue
        }.toMap()
    }
}

File: packages/davinci/ios/Collector/DaVinciCollectorValueApplier.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingDavinci
import PingDavinciPlugin
import PingOrchestrate

/// Applies collector value mutations from JS payloads to active DaVinci collectors.
///
/// DaVinci collectors are identified by key (`collector.id`), not by type index.
enum DaVinciCollectorValueApplier {
  /// Result indicating whether the applied collector triggers an immediate form submit.
  struct ApplyResult: Sendable {
    /// True when a FlowCollector key was applied as part of the input.
    let isFlowTrigger: Bool
  }

  /// Parsed collector mutation instruction received from JavaScript.
  ///
  /// - Note: `@unchecked Sendable` is used because `value` is `Any?`, which is not
  ///   statically sendable. Parsed mutations are immutable value types and consumed
  ///   synchronously within native mutation flow.
  struct CollectorMutation: @unchecked Sendable {
    let key: String
    let value: Any?
  }

  /// Parses collector mutation payload from bridge input.
  ///
  /// The input format matches `DaVinciNextInput`:
  /// `{ collectors: [{ key: string, value: unknown }] }`.
  ///
  /// - Parameter input: Bridge input map.
  /// - Returns: Parsed collector mutations.
  /// - Throws: `DaVinciBridgeError.argument` when payload is malformed.
  static func parseInput(_ input: NSDictionary) throws -> [CollectorMutation] {
    guard let collectors = input["collectors"] as? [Any] else {
      return []
    }

    var mutations = [CollectorMutation]()
    for (index, item) in collectors.enumerated() {
      guard let entry = item as? NSDictionary else {
        throw DaVinciBridgeError.argument("Invalid collector entry at index \(index)")
      }
      guard let key = entry["key"] as? String, !key.isEmpty else {
        throw DaVinciBridgeError.argument("Missing 'key' in collector entry at index \(index)")
      }
      mutations.append(
        CollectorMutation(
          key: key,
          value: readDynamicValue(entry["value"])
        )
      )
    }
    return mutations
  }

  /// Applies parsed collector mutations to an active continue node.
  ///
  /// - Parameters:
  ///   - continueNode: Active continue node providing the collector instances.
  ///   - mutations: Parsed collector mutations.
  /// - Returns: Result indicating whether a FlowCollector key was applied.
  /// - Throws: `DaVinciBridgeError` when mutation fails.
  static func apply(
    _ continueNode: ContinueNode,
    mutations: [CollectorMutation]
  ) throws -> ApplyResult {
    let collectorsByKey = Dictionary(
      uniqueKeysWithValues: continueNode.collectors.map { ($0.id, $0) }
    )

    var isFlowTrigger = false
    for mutation in mutations {
      guard let collector = collectorsByKey[mutation.key] else {
        throw DaVinciBridgeError.argument("No active collector found for key='\(mutation.key)'")
      }
      try applyValue(collector, key: mutation.key, value: mutation.value)
      if collector is FlowCollector {
        isFlowTrigger = true
      }
    }
    return ApplyResult(isFlowTrigger: isFlowTrigger)
  }

  /// Apply a single value to a collector instance.
  ///
  /// - Parameters:
  ///   - collector: Target collector.
  ///   - key: Collector key (used for error messages).
  ///   - value: Value from JS.
  /// - Throws: `DaVinciBridgeError` when the value type is incompatible with the collector.
  private static func applyValue(_ collector: any Collector, key: String, value: Any?) throws {
    switch collector {
    case let multiSelectCollector as MultiSelectCollector:
      let list = try asStringList(value, key: key)
      multiSelectCollector.value = list

    case let phoneNumberCollector as PhoneNumberCollector:
      let map = asStringMap(value)
      if let countryCode = map["countryCode"] {
        phoneNumberCollector.countryCode = countryCode
      }
      if let phoneNumber = map["phoneNumber"] {
        phoneNumberCollector.phoneNumber = phoneNumber
      }

    case let registrationCollector as DeviceRegistrationCollector:
      let map = asStringMap(value)
      guard let deviceType = map["type"] else {
        throw DaVinciBridgeError.argument(
          "DeviceRegistrationCollector key='\(key)': value map must include 'type'"
        )
      }
      guard let device = registrationCollector.devices.first(where: { $0.type == deviceType }) else {
        throw DaVinciBridgeError.argument(
          "DeviceRegistrationCollector key='\(key)': no device found with type='\(deviceType)'"
        )
      }
      registrationCollector.value = device

    case let authenticationCollector as DeviceAuthenticationCollector:
      let map = asStringMap(value)
      guard let deviceType = map["type"] else {
        throw DaVinciBridgeError.argument(
          "DeviceAuthenticationCollector key='\(key)': value map must include 'type'"
        )
      }
      // Fall back to constructing a Device from the map when the type is not in the
      // known device list — mirrors Android DaVinciCollectorValueApplier behaviour.
      let device: Device
      if let known = authenticationCollector.devices.first(where: { $0.type == deviceType }) {
        device = known
      } else {
        device = try makeFallbackDevice(from: map, type: deviceType, key: key)
      }
      authenticationCollector.value = device

    default:
      // All remaining collectors (TextCollector, PasswordCollector, SingleSelectCollector,
      // SubmitCollector, FlowCollector) extend SingleValueCollector which accepts a String
      // via initialize(with:).
      guard collector is SingleValueCollector else {
        throw DaVinciBridgeError.unsupportedCollector(
          "Collector key='\(key)' type=\(String(describing: type(of: collector))) is not supported by the bridge"
        )
      }
      let stringValue = try asString(value, key: key)
      collector.initialize(with: stringValue)
    }
  }

  // MARK: - Value coercion helpers

  /// Read a dynamic value from a bridge entry.
  ///
  /// - Parameter value: Raw bridge value.
  /// - Returns: Dynamic value, or `nil` when null.
  private static func readDynamicValue(_ value: Any?) -> Any? {
    if value is NSNull {
      return nil
    }
    return value
  }

  /// Coerce a dynamic value to a string.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - key: Collector key (used in error messages).
  /// - Returns: String value.
  /// - Throws: `DaVinciBridgeError.argument` when value cannot be represented as a string.
  private static func asString(_ value: Any?, key: String) throws -> String {
    switch value {
    case let string as String:
      return string
    case let number as NSNumber:
      // CFBooleanGetTypeID() distinguishes an actual Swift/ObjC Bool from a bridged
      // integer NSNumber — both satisfy `as? Bool` in Swift, so we must check the
      // CF type first to avoid converting numeric 0/1 to "false"/"true".
      if CFGetTypeID(number) == CFBooleanGetTypeID() {
        return number.boolValue ? "true" : "false"
      }
      return number.stringValue
    default:
      throw DaVinciBridgeError.argument(
        "Collector key='\(key)' expects a string-compatible value"
      )
    }
  }

  /// Coerce a dynamic value to a list of strings.
  ///
  /// - Parameters:
  ///   - value: Dynamic value.
  ///   - key: Collector key (used in error messages).
  /// - Returns: Array of strings.
  /// - Throws: `DaVinciBridgeError.argument` when value is not an array or contains a non-String element.
  private static func asStringList(_ value: Any?, key: String) throws -> [String] {
    let elements: [Any]
    if let array = value as? [Any] {
      elements = array
    } else if let array = value as? NSArray {
      elements = array as [Any]
    } else {
      throw DaVinciBridgeError.argument(
        "Collector key='\(key)' expects an array of strings"
      )
    }
    return try elements.map { element in
      guard let string = element as? String else {
        throw DaVinciBridgeError.argument(
          "Collector key='\(key)' expects an array of strings, got \(type(of: element))"
        )
      }
      return string
    }
  }

  /// Construct a `Device` from a string map when the type is absent from the collector's
  /// known device list. Mirrors the Android bridge fallback in `DaVinciCollectorValueApplier`.
  ///
  /// - Parameters:
  ///   - map: String map extracted from the JS value.
  ///   - type: Resolved device type string.
  ///   - key: Collector key (used for error messages).
  /// - Returns: A `Device` decoded from the map fields.
  /// - Throws: `DaVinciBridgeError.argument` when the JSON cannot be decoded.
  private static func makeFallbackDevice(
    from map: [String: String],
    type deviceType: String,
    key: String
  ) throws -> Device {
    var json: [String: Any] = [
      "type": deviceType,
      "title": map["title"] ?? deviceType,
      "iconSrc": map["iconSrc"] ?? "about:blank",
      "default": false
    ]
    if let id = map["id"] { json["id"] = id }
    if let description = map["description"] { json["description"] = description }
    do {
      let data = try JSONSerialization.data(withJSONObject: json)
      return try JSONDecoder().decode(Device.self, from: data)
    } catch {
      throw DaVinciBridgeError.argument(
        "DeviceAuthenticationCollector key='\(key)': failed to construct fallback device"
      )
    }
  }

  /// Coerce a dynamic value to a string-keyed string map.
  ///
  /// - Parameter value: Dynamic value.
  /// - Returns: String map.
  private static func asStringMap(_ value: Any?) -> [String: String] {
    guard let dictionary = value as? NSDictionary else {
      return [:]
    }
    var mapped = [String: String]()
    for (rawKey, rawValue) in dictionary {
      guard let key = rawKey as? String else { continue }
      if let string = rawValue as? String {
        mapped[key] = string
      } else if let number = rawValue as? NSNumber {
        mapped[key] = number.stringValue
      }
    }
    return mapped
  }
}
